### DISCUSSION TITLE
state#update interferes with with page fill

### DISCUSSION THREAD

**Discussion by muekoeff** [typst/typst#4073] [General]
### Description

When updating some `state` using `.update()` subsequent `page(fill: …)` are ignored.

#### Reproduce
1. Simple example:
    ```
    #let some_state = state("some_state")
    #let some_theme(body) = {
        some_state.update(true)
        body
    }
    
    #show: some_theme.with()
    #set page(fill: blue)
    #pagebreak()
    #set page(fill: white)
    ```
    I would expect the first page to be blue and the second to be white.
    Instead, both pages are white ⇒ [output.pdf](https://github.com/typst/typst/files/15214604/test.pdf)
2. Setting page fill before state#update:
    ```
    #let some_state = state("some_state")
    #let some_theme(body) = {
        set page(fill: blue)
        some_state.update(true)
        body
    }
    
    #show: some_theme.with()
    #set page(fill: white)
    #pagebreak()
    #set page(fill: blue)
    ```
    I would expect the first page to be white and the second to be blue.
    Instead, both pages are blue ⇒ [output.pdf](https://github.com/typst/typst/files/15214609/test.pdf)

#### Context
I encountered this issue when creating a theme for Polylux. I want to use `state`s to store user preferences but would like to still be able to set individual slides' background colours, for example for title slides:
```
#import "@preview/polylux:0.3.1": logic

#let accent-colour = state("accent-colour")

#let some-theme(colour: none, body) = {
    accent-colour.update(colour)
    body
}
#let title-slide(body) = {
    context {
        set page(fill: accent-colour.get())
    }
    logic.polylux-slide(body)
}
#let slide(body) = {
    set page(fill: white)
    logic.polylux-slide(body)
}

#show: some-theme.with(colour: blue)
#title-slide[Lorem]
#slide[Ipsum]
```

`typst -V`: `typst 0.11.0 (2bf9f95d)`

### Operating system

Linux

### Typst version

- [X] I am using the latest version of Typst

**Comment by Enivex** [comment]
Unless the page is empty, the `set page` will only apply from the next page. The first page isn't empty, because of the content returned by the state update.

Since there is another set page right after the `pagebreak`, the one right before it will never be applied to any page.

**Reply by Enivex** [reply]
If you do that, then the update won't happen. The content is necessary, as explained in the documentation for update 

Edit: I think the comment I replied to was deleted.

**Reply by muekoeff** [reply]
Noticed my mistake and deleted my comment with the hope that you hadn't read it yet.

Original comment:
> Writing `let _ = some_state.update(true)` instead achieved what I wanted. Thank you very much for the explanation (and sorry for the the wrong bug report).

**Reply by Enivex** [reply]
If you need help, you may want to check out the Discord, by the way.
